sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -42,10 +42,34 @@ case class InMemoryTableScanExec(
   override def output: Seq[Attribute] = attributes
 
   // The cached version does not change the outputPartitioning of the original SparkPlan.
-  override def outputPartitioning: Partitioning = relation.child.outputPartitioning
+  // But the cached version could alias output, so we need to replace output.
+  override def outputPartitioning: Partitioning = {
```

Comment:
I suppose you mean to keep the change of `InMemoryTableScanExec`. Other changes are removed. Can you take a look? Thanks.